pkg/frontend/querymiddleware/cardinality_test.go

```diff
@@ -0,0 +1,341 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/dskit/cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/user"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/util"
+)
+
+func Test_cardinalityEstimateBucket_GenerateCacheKey_keyFormat(t *testing.T) {
+	requestTime := parseTimeRFC3339(t, "2023-01-09T03:24:12Z")
+	hoursSinceEpoch := util.TimeToMillis(requestTime) / time.Hour.Milliseconds()
+	daysSinceEpoch := hoursSinceEpoch / 24
+
+	type args struct {
+		userID string
+		r      Request
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
```

Comment:
[nit] We're typically used to call it "expected" instead of "want". Just a nit, if you want to be consistent.